plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2398,6 +2414,99 @@ else if (isMetadataColumnId(columnHandle.getId())) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<ConnectorExpression> projections = ImmutableList.builder();
+        ImmutableList.Builder<Assignment> resultAssignments = ImmutableList.builder();
+        ImmutableList.Builder<IcebergColumnHandle> aggregateColumnsBuilder = ImmutableList.builder();
+
+        Set<IcebergColumnHandle> projectionsSet = new HashSet<>();
+
+        if (aggregates.size() != 1) {
+            // not handling multiple aggregations for now
+            return Optional.empty();
+        }
+
+        AggregateFunction aggregate = aggregates.get(0);
+
+        Optional<AggregateExpression> rewriteResult = aggregateFunctionRewriter.rewrite(session, aggregate, assignments);
+        if (rewriteResult.isEmpty()) {
+            return Optional.empty();
+        }
+        AggregateExpression aggregateExpression = rewriteResult.get();
+
+        if (aggregateExpression.getFunction().startsWith("count")) {
+            IcebergColumnHandle aggregateIcebergColumnHandle = new IcebergColumnHandle(aggregateExpression.toColumnIdentity(AggregateExpression.COUNT_AGGREGATE_COLUMN_ID),
+                    aggregate.getOutputType(), ImmutableList.of(), aggregate.getOutputType(), Optional.empty());
+            aggregateColumnsBuilder.add(aggregateIcebergColumnHandle);
+            projections.add(new Variable(aggregateIcebergColumnHandle.getName(), aggregateIcebergColumnHandle.getType()));
+            projectionsSet.add(aggregateIcebergColumnHandle);
+            resultAssignments.add(new Assignment(aggregateIcebergColumnHandle.getName(), aggregateIcebergColumnHandle, aggregateIcebergColumnHandle.getType()));
+        }
+
+        IcebergTableHandle tableHandleTemp = new IcebergTableHandle(tableHandle.getSchemaName(),
+                tableHandle.getTableName(),
+                tableHandle.getTableType(),
+                tableHandle.getSnapshotId(),
+                tableHandle.getTableSchemaJson(),
+                tableHandle.getPartitionSpecJson(),
+                tableHandle.getFormatVersion(),
+                tableHandle.getUnenforcedPredicate(),
+                tableHandle.getEnforcedPredicate(),
+                tableHandle.getLimit(),
+                projectionsSet,
+                tableHandle.getNameMappingJson(),
+                tableHandle.getTableLocation(),
+                tableHandle.getStorageProperties(),
+                tableHandle.isRecordScannedFiles(),
+                tableHandle.getMaxScannedFileSize());
+
+        return Optional.of(new AggregationApplicationResult<>(tableHandleTemp, projections.build(), resultAssignments.build(), ImmutableMap.of(), false));
+    }
+
+    private boolean hasEqualityDeletes(ConnectorSession session, IcebergTableHandle tableHandle)
+    {
+        Table icebergTable = catalog.loadTable(session, tableHandle.getSchemaTableName());
+
+        if (icebergTable.currentSnapshot().summary().containsKey(SnapshotSummary.TOTAL_EQ_DELETES_PROP)) {
```

Comment:
Is the summary guaranteed to contain the `total-equality-deletes` if equality deletes exists?

i can't find it on https://iceberg.apache.org/spec/#snapshots

Reply:
You are correct, looks like Snapshot's summary is not updated in the spec.

Though it is being update [here](https://github.com/apache/iceberg/blob/aa891acf20040d15e7ca59dc503adb3c1e4325b8/core/src/main/java/org/apache/iceberg/metrics/CommitMetricsResult.java#L165).

Seems to be this metric will be updated by equality delete operation, and we can use it.

Added a test-case to show with `equalityDelete` w/o `aggregatePushDown`  count queries are good.

Reply:
I don't think I'd use anything that's not spec'd out as a source of truth for determining if equality deletes exist.
The reason is some other engine/process can technically write equality deletes but not write this summary property and then Trino would pass the check when it should not (potentially leading to correctness issues).

Beyond that there are other arguments to not using this snapshot summary property such as , what if none of the equality deletes even apply for a query with a given predicate? I think we'd still want to do the aggregation push down because it's safe to do so. However, with just a property > 0 check we'd say aggregation pushdown is not possible.

In the Spark Iceberg integration we actually do a scan planning, go through tasks and if a task has *any* deletes, we determine that we can't move forward with agg pushdown. This addresses this concern but I'm not sure yet how that would look like in the Trino model. I think the splits would need to plumb through this information?